src/renderers/shared/fiber/ReactFiberBeginWork.js

```diff
@@ -54,17 +54,35 @@ module.exports = function<T, P, I, C>(config : HostConfig<T, P, I, C>) {
     workInProgress.pendingWorkPriority = NoWork;
   }
 
-  function updateClassComponent(current, workInProgress) {
+  function updateClassComponent(current : ?Fiber, workInProgress : Fiber) {
     var props = workInProgress.pendingProps;
-    if (!workInProgress.stateNode) {
+    var instance = workInProgress.stateNode;
+    if (!instance) {
       var ctor = workInProgress.type;
-      workInProgress.stateNode = new ctor(props);
+      workInProgress.stateNode = instance = new ctor(props);
+    } else if (typeof instance.shouldComponentUpdate === 'function') {
+      if (current && current.memoizedProps) {
+        // Revert to the last flushed props, incase we aborted an update.
+        instance.props = current.memoizedProps;
+        if (!instance.shouldComponentUpdate(props)) {
+          return bailoutOnCurrent(current, workInProgress);
+        }
+      }
+      if (!workInProgress.hasWorkInProgress && workInProgress.memoizedProps) {
```

Comment:
I meant exhaustive for the updates case I guess. My initial thought would be that `current && current.memoizedProps` would always be set when doing an update.

In this code, how do we distinguish the ping-pong case from other kinds of updates? I am having trouble thinking about this code because I would have probably expected some sort of `===` check on the props.